src/category_theory/limits/kan_extension.lean

```diff
@@ -0,0 +1,370 @@
+def diagram (F : S ⥤ D) (x : L) : index ι x ⥤ D :=
+  comma.fst ι (functor.from_punit x) ⋙ F
+variable {ι}
+
+/-- A cocone over `Lan.diagram ι F x` used to define `Lan`. -/
+@[simp]
+def cocone {F : S ⥤ D} {G : L ⥤ D} (x : L) (f : F ⟶ ι ⋙ G) :
+  cocone (diagram ι F x) :=
+{ X := G.obj x,
+  ι :=
+  { app := λ i, f.app i.left ≫ G.map i.hom,
+    naturality' := begin
+      rintro ⟨ir,⟨il⟩,i⟩ ⟨jl,⟨jr⟩,j⟩ ⟨fl,⟨⟨fl⟩⟩,ff⟩,
+      dsimp at *,
+      simp only [functor.comp_map, category.comp_id, nat_trans.naturality_assoc],
+      rw [← G.map_comp, ff],
+      tidy,
+    end } }
+
+variable (ι)
+
+/-- An auxiliary definition used to define `Lan`. -/
+@[simps]
+def obj_aux (F : S ⥤ D) [∀ x, has_colimits_of_shape (index ι x) D] : L ⥤ D :=
+{ obj := λ x, colimit (diagram ι F x),
+  map := λ x y f, colimit.pre (diagram _ _ _) (index.map f),
+  map_id' := begin
+    intro l,
+    ext j,
+    erw [colimit.ι_pre, category.comp_id],
+    congr' 1,
+    rw index.map_id,
+  end,
+  map_comp' := begin
+    intros x y z f g,
+    ext j,
+    have := colimit.pre_pre (diagram ι F z) (index.map g) (index.map f),
+    change _ = _ ≫
+      colimit.pre (index.map g ⋙ diagram ι F z) (index.map f) ≫
+      colimit.pre (diagram ι F z) (index.map g),
+    rw this,
+    change _ = colimit.ι ((index.map f ⋙ index.map g) ⋙ diagram ι F z) j ≫ _,
+    rw [colimit.ι_pre, colimit.ι_pre],
+    congr' 1,
+    simp only [index.map_comp, functor.comp_obj],
+  end }
+
+/-- An auxiliary definition used to define `Lan` and `Lan.adjunction`. -/
+@[simps]
+def equiv [∀ x, has_colimits_of_shape (index ι x) D] (F : S ⥤ D) (G : L ⥤ D) :
+  (obj_aux ι F ⟶ G) ≃ (F ⟶ ι ⋙ G ) :=
+{ to_fun := λ f,
+  { app := λ x, by apply colimit.ι (diagram ι F (ι.obj x)) (index.mk (𝟙 _)) ≫ f.app _, -- sigh
+  naturality' := begin
+    intros x y ff,
+    simp,
+    erw [← f.naturality (ι.map ff)],
+    delta obj_aux,
+    erw [← category.assoc, ← category.assoc],
+    erw colimit.ι_pre (diagram ι F (ι.obj y)) (index.map (ι.map ff)) (index.mk (𝟙 _)),
+    congr' 1,
+    have := colimit.w (diagram ι F (ι.obj y)) (index.mk_hom (𝟙 _) ff),
+    convert this,
+    rw index.map_mk,
+    simp [index.map_mk],
+  end },
+  inv_fun := λ f,
+  { app := λ x, colimit.desc (diagram ι F x) (cocone _ f),
+    naturality' := begin
+      intros x y ff,
+      ext j,
+      erw [colimit.pre_desc, ← category.assoc, colimit.ι_desc, colimit.ι_desc],
+      tidy,
+    end },
+  left_inv := begin
+    intro x,
+    ext k j,
+    rw colimit.ι_desc,
+    dsimp only [cocone],
+    rw [category.assoc, ← x.naturality j.hom, ← category.assoc],
+    congr' 1,
+    dsimp only [obj_aux, index.map],
+    change colimit.ι _ _ ≫ colimit.pre (diagram ι F k) (index.map _) = _,
+    rw colimit.ι_pre,
+    congr,
+    cases j,
+    tidy,
+  end,
+  right_inv := by tidy }
+
+/-- A variant of `Lan.equiv` with `whiskering_left` instead of functor composition. -/
+@[simps]
+def equiv' [∀ x, has_colimits_of_shape (index ι x) D] (F : S ⥤ D) (G : L ⥤ D) :
```

Comment:
I'm guessing you realised (just like I did) that it should be written in this form so that `adjunction.left_adjoint_of_equiv` knows whats going on - but perhaps this is a sign that `adjunction.left_adjoint_of_equiv` should have the right adjoint as an explicit argument

Reply:
I agree with this, but I think this should be something for a separate PR as this will likely required changing many other unrelated files.